### ISSUE TITLE
Quality of generated content is drastically lower with max_new_tokens set at max

### ISSUE THREAD

**Issue by michael-imbeault** [oobabooga/text-generation-webui#2246]
### Describe the bug

If you set max_new_tokens at the maximum (2000), the quality of the generated content is way lower, and the models hallucinate a lot more. 

Seen with multiple models / questions - I`m using mostly 13B and 30B Llama variants quantized to 4 bits.

Setting it at the max level is a natural reflex for new users, to get long / more complex answers - this bug might make some new users think that models / oobabooga are less performant than what they are.

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

Take an article - let`s say https://edition.cnn.com/2023/05/20/europe/russia-warning-f-16-jets-ukraine-intl/index.html

Here`s what I input in the model:

> Russia’s deputy foreign minister has warned Western countries of “enormous risks” if Ukraine is provided with F-16 fighter jets, Russian state media TASS reported Saturday.
> 
> The comments come after US President Joe Biden gave his backing for Ukrainian pilots to be trained to fly F-16s, reversing his previous position.
> 
> F-16s are considered high performance weapon systems with a range of 500 miles (860 kilometers), and would be an upgrade to the aircraft currently in Ukraine’s fleet.
> 
> Responding to the move, Alexander Grushko said: “We see that the Western countries are still adhering to the escalation scenario.
> 
> “It involves enormous risks for themselves. In any case, this will be taken into account in all our plans, and we have all the necessary means to achieve the set goals.”
> 
> President Joe Biden walks alongside British Prime Minister Rishi Sunak, European Commission President Ursula von der Leyen, Canadian Prime Minister Justin Trudeau, Japan's Prime Minister Fumio Kishida, Italian Premier Giorgia Meloni, European Council President Charles Michel, French President Emmanuel Macron and German Chancellor Olaf Scholz to a flower wreath laying ceremony at the Hiroshima Peace Memorial Park in Japan on Friday.
> 
> Biden tells G7 allies the US will support joint F-16 training effort for Ukrainians
> 
> Earlier this week, the UK and the Netherlands were reportedly building an “international coalition” to help Ukraine procure F-16 fighter jets.
> 
> A handful of European countries have a supply of US-made F-16s, including the Netherlands, which has signaled a willingness to export some of them to Ukraine. But the US would have to approve that third-party transfer because of the jets’ sensitive US technology.
> 
> Since Russia’s invasion started, Ukrainian President Volodymyr Zelensky, has struggled to get F-16s to aid his fight, over fears they could be used on Russian soil, potentially triggering an escalation between NATO and Russia.
> 
> After saying Kyiv did not need the fighter jets earlier this year, US President Joe Biden reversed his objections by signaling to European allies that they would allow F-16 exports to Ukraine.
> 
> Biden’s top national security adviser, Jake Sullivan, told reporters Saturday in Japan Biden believes in equipping the country for a long-term fight against Russia.
> 
> In March, the US hosted two Ukrainian pilots at a military base in Tucson, Arizona, to evaluate their skills using flight simulators and assess how long it would take them to learn to fly various US military aircraft, including F-16s.
> 
> Factually summarize the above in 5 bullet points:

Here's the answer at max_new_tokens 1024 - it`s great:

>     The G7 leaders visited the Hiroshima Peace Memorial Park in Japan on Friday for a flower wreath laying ceremony.
> 
>     The United States is supporting a joint F-16 training effort for Ukrainian pilots with its G7 allies.
> 
>     The UK and the Netherlands are reportedly building an "international coalition" to assist Ukraine in procuring F-16 fighter jets.
> 
>     The US must approve any third-party transfer of F-16s due to the jets' sensitive US technology.
> 
>     After initially opposing the idea, President Joe Biden has now indicated his support for equipping Ukraine with F-16s to aid in their fight against Russia.

Here's the answer at max_new_tokens 1960 - still great:

>     The United States is considering providing Ukraine with advanced fighter jets such as the F-35 or F/A-18E/F Super Hornet.
> 
>     This move comes after President Biden's decision to provide Javelin anti-tank missiles to Ukraine.
> 
>     The goal of this potential transfer is to increase Ukraine's air defense capabilities and deter Russian aggression.
> 
>     The US has already provided training and equipment to Ukraine, but the provision of advanced fighter jets would be a significant escalation.
> 
>     The evaluation of Ukrainian pilots at a military base in Tucson, Arizona, was aimed at determining their ability to operate US military aircraft and the time required for them to become proficient in flying these planes.

However, if you set max_new_tokens to 2000, it hallucinates completely:

>     The AI language model GPT-3 was trained on a diverse range of internet text data.
> 
>     It has been shown to be capable of generating human-like responses to questions related to aviation and flight training.
> 
>     However, there is no specific information available about its knowledge or capabilities regarding simulating realistic flight scenarios for pilots using advanced simulation software.
> 
>     Simulation software such as DCS World can provide highly detailed and accurate representations of aircraft performance and handling characteristics.
> 
>     It may require significant time and effort for an individual with little or no prior experience flying to become proficient at operating complex military aircraft like the F-16 through simulation alone.

### Logs

```shell
No logs
```

### System Info

```shell
3090 RTX
MetaIX_GPT4-X-Alpasta-30b-4bit
Llama-precise settings
```

**Comment by Ph0rk0z** [comment]
You basically reduce the context to 48 so the model has nothing to go off of.

**Comment by michael-imbeault** [comment]
That is not how it works - the initial context contains the full prompt, then it starts generating from there - I'm not having multiple inputs here, just one - also it works great at any chosen max_new_tokens (i.e. 1960) that is not the maximum.

**Comment by Ph0rk0z** [comment]
Your question is probably truncated too if it didn't fit in 48 tokens. 2000 means the max prompt size is 48.

**Comment by michael-imbeault** [comment]
Yes I get it now - could we have a little blurb of text on max_new_tokens in the settings section so this is better understood by users? My initial thoughts were that context was fully separate from the answer it generates, but now I get its initial context + max_new_tokens that need to be below maximum prompt size.

Notice that that's where the confusion comes from - the next parameter is 'maximum prompt size', which led me to believe it was separate from the new tokens in terms of context.

Any chance as well we have some indication of number of tokens (see previous issue I opened) in the GUI? would be nice if it is red / shows a warning when full prompt size + max new tokens exceeds the limit.

**Comment by CySlider** [comment]
I am also completely confused by this setting, and fell into the same trap and would like to get a better understanding of this setting.

So here is what I intuitively thought. 

max_new_tokens = how many tokens the LLM is allowed to generate in it's reply before stopping early. 

But increasing it, reduces the input context. so the higher the max_new_token value, the less of the prompt he uses as context, explaining weird behavior when it gets too low.

I don't understand why one influences the other. I can obviously chat with a bot much longer than 2000 tokens context just fine. (yes older parts of the conversation will be omitted, but I still have a long context window and the overall conversation does not stop when reaching 2000 tokens) 

 Why can't I have a big context and still allow it to generate 2000 or even 3000 new tokens?

**Comment by CySlider** [comment]
Ok, I think I start to get it. If I am correct then the following is the case:

In a single "low level" prompt, He can only generate text up to his full context length, including the prompt itself
So if I have a 500 token prompt, the maximum "space" to generate a response is ~1500 tokens on a 2000 token prompt window model

If I set max_new_tokens high, the prompt gets truncated to make it not use more than 2000 - max_new_tokens tokens  of this space, so the LMM is guaranteed to have space for his answer, even if he will not need that much space

If you want him to have a lot of context so he can understand more, you need to set a very low max_new_tokens but he will more often stop early in that case.

You can press continue, so he will generate a new chunk of max_new_tokens, cutting the prompt if necessary to accommodate the new text.

Example:

```
max context window = 2000
max_new_tokens = 100
prompt_size = 1800

Answer: 0 - 100 tokens

If 100 tokens are used, and continue is pressed

prompt_size = 1900

Answer: 0 - 100 tokens

If 100 tokens are used, and continue is pressed

prompt gets truncated ( I guess in the beginning and as it seems whole sentences or lines are being removed=

prompt_size = 1900  (or lower again)

Answer: 0 - 100 tokens
```

So it seems it is really advisable to use a low max_new_tokens value to increase the quality of the response if you have a big context (like a long chat)

To not have to press continue all over again, there is actually a useful settings to automate it in chat mode: 

`Generation attempts (for longer replies)`

It will press continue for you for up to that many times if generated tokens are at the limit. 

I think this is a very very crucial part of the model behavior that needs to be communicated WAY BETTER in the UI.
At least show the max prompt length value when modifying `max_new_tokens` to give the user a hint of what he is playing with here.
